From ee8327d46875a5f08a0a4698e874034d669b19ff Mon Sep 17 00:00:00 2001
From: Aarni Koskela <akx@iki.fi>
Date: Wed, 28 Apr 2021 10:33:40 +0300
Subject: [PATCH 1/2] Run locale identifiers through `os.path.basename()`

---
 babel/localedata.py      |  2 ++
 tests/test_localedata.py | 31 ++++++++++++++++++++++++++++++-
 2 files changed, 32 insertions(+), 1 deletion(-)

diff --git a/babel/localedata.py b/babel/localedata.py
index a638e58..76e2195 100644
--- a/babel/localedata.py
+++ b/babel/localedata.py
@@ -48,6 +48,7 @@ def exists(name):
     """
     if not name or not isinstance(name, string_types):
         return False
+    name = os.path.basename(name)
     if name in _cache:
         return True
     file_found = os.path.exists(os.path.join(_dirname, '%s.dat' % name))
@@ -92,6 +93,7 @@ def load(name, merge_inherited=True):
     :raise `IOError`: if no locale data file is found for the given locale
                       identifer, or one of the locales it inherits from
     """
+    name = os.path.basename(name)
     _cache_lock.acquire()
     try:
         data = _cache.get(name)
diff --git a/tests/test_localedata.py b/tests/test_localedata.py
index 37c1304..555e06f 100644
--- a/tests/test_localedata.py
+++ b/tests/test_localedata.py
@@ -11,11 +11,18 @@
 # individuals. For the exact contribution history, see the revision
 # history and logs, available at http://babel.edgewall.org/log/.
 
+import os
+import pickle
+import sys
+import tempfile
 import unittest
 import random
 from operator import methodcaller
 
-from babel import localedata, numbers
+import pytest
+
+from babel import localedata, Locale, UnknownLocaleError
+
 
 class MergeResolveTestCase(unittest.TestCase):
 
@@ -105,3 +112,25 @@ def test_locale_argument_acceptance():
     assert normalized_locale is None
     locale_exist = localedata.exists(['en_us', None])
     assert locale_exist == False
+
+
+def test_locale_name_cleanup():
+    """
+    Test that locale identifiers are cleaned up to avoid directory traversal.
+    """
+    no_exist_name = os.path.join(tempfile.gettempdir(), "babel%d.dat" % random.randint(1, 99999))
+    with open(no_exist_name, "wb") as f:
+        pickle.dump({}, f)
+
+    try:
+        name = os.path.splitext(os.path.relpath(no_exist_name, localedata._dirname))[0]
+    except ValueError:
+        if sys.platform == "win32":
+            pytest.skip("unable to form relpath")
+        raise
+
+    assert not localedata.exists(name)
+    with pytest.raises(IOError):
+        localedata.load(name)
+    with pytest.raises(UnknownLocaleError):
+        Locale(name)
\ No newline at end of file
-- 
2.17.1

